ctapipe/core/tests/test_qualityquery.py

```diff
@@ -126,22 +140,87 @@ def test_table_mask():
     np.testing.assert_equal(stats["cumulative_counts"], [5, 3, 1])
 
 
-def test_printing():
+def test_to_table_after_call(subarray_prod5_paranal):
+    """Test counting and conversion to table."""
+    query = QualityQuery(
+        quality_criteria=[
+            ("inside unit circle", [("type", "*", "(x**2 + y**2) < 1.0")]),
+            ("less than half", [("type", "*", "x < 0.5")]),
+        ],
+        subarray=subarray_prod5_paranal,
+    )
+    assert np.all(query(tel_id=1, x=0.3, y=0))
+    assert not np.all(query(tel_id=1, x=1, y=0))
+
+    assert np.all(query(tel_id=2, x=0.3, y=0))
+    assert not np.all(query(tel_id=2, x=1, y=0))
+
+    stats = query.to_table()
+    np.testing.assert_equal(stats["counts"], [4, 2, 2])
+    np.testing.assert_equal(stats["cumulative_counts"], [4, 2, 2])
+
+
+def test_printing(subarray_prod5_paranal):
     """Just check the query can be stringified correctly"""
     query = QualityQuery(
-        quality_criteria=[("check", "x>3")],
+        quality_criteria=[("check", [("type", "*", "x>3")])],
+        subarray=subarray_prod5_paranal,
     )
 
     assert isinstance(str(query), str)
 
 
-def test_setup():
+def test_setup(subarray_prod5_paranal):
     """Test that tuples can only have length 2"""
-    from traitlets import TraitError
+    query = QualityQuery(
+        quality_criteria=[
+            ("foo", [("type", "*", "x > 0")]),
+            ("bar", [("type", "*", "x > 1"), ("id", "1", "x > 2")]),
+            ("baz", "x > 3"),
+        ],
+        subarray=subarray_prod5_paranal,
+    )
+
+    query(tel_id=1, x=0)
 
-    # 2-tuple works
-    QualityQuery(quality_criteria=[("1", "2")])
 
-    # 3-tuple fails
-    with pytest.raises(TraitError):
-        QualityQuery(quality_criteria=[("1", "2", "3")])
+def test_with_lambda(subarray_prod5_paranal):
+    """Test that we raise an error when using lambda expresssions,
+    which we don't support since v0.16.0"""
+
+    with pytest.raises(ValueError, match="lambda"):
+        QualityQuery(
+            quality_criteria=[("name", [("type", "*", "lambda p: p > 2")])],
+            subarray=subarray_prod5_paranal,
+        )
+
+
+def test_telescope_component(subarray_prod5_paranal):
+    """Test that more specific patterns override more general."""
+    query = QualityQuery(
+        quality_criteria=[
+            ("foo", [("type", "*", "x > 0"), ("id", "2", "x < 0")]),
+        ],
+        subarray=subarray_prod5_paranal,
+    )
+
+    x = 1
+    assert query(x=x, tel_id=1)
+    assert not query(x=x, tel_id=2)
```

Comment:
Shouldn't this return a tuple? I.e. wouldn't you need to test the single elements?

Reply:
array with single item: `foo`